### PR TITLE
Refactor weights and objective fns to get comm at setup

### PR DESCRIPTION
**Pull request by bvanessen**
Updated both the weights an objective functions to get their comm object pointers after construction.  This is intended to make it easier to serialize and deserialize these classes with Cereal.

### CODE REVIEW

File: include/lbann/weights/weights.hpp [hunk 1]
```diff
@@ -77,6 +77,8 @@ class weights {
   /** Get weights name. */
   std::string get_name() const { return m_name; }
 
+  lbann_comm* get_comm() const { return m_comm; }
```
Comment (timmoon10) [inline include/lbann/weights/weights.hpp]: We should return a reference. If `m_comm` is a null pointer, we should throw an exception.

File: include/lbann/weights/weights.hpp [hunk 2]
```diff
@@ -165,6 +167,11 @@ class weights {
   El::DistData get_matrix_distribution() const;
   void set_matrix_distribution(El::DistData dist);
 
+  // -----------------------------------------------
+  // Initialize the weight class
+  // -----------------------------------------------
+  void init(lbann_comm& comm);
```
Comment (timmoon10) [inline include/lbann/weights/weights.hpp]: It's not clear what the difference is between `init` and `setup`. We should call it `set_comm`.

File: include/lbann/weights/weights.hpp (hunk 1 above)
Comment (bvanessen) [reply include/lbann/weights/weights.hpp]: Done.

File: include/lbann/weights/weights.hpp (hunk 2 above)
Comment (bvanessen) [reply include/lbann/weights/weights.hpp]: Good idea. Done.

File: include/lbann/weights/weights.hpp [hunk 3]
```diff
@@ -63,6 +63,13 @@ class optimizer;
  */
 class weights {
   friend class optimizer;
+private:
+  weights();
+  // -----------------------------------------------
+  // Internal method for setting the comm pointer
+  // -----------------------------------------------
+  void set_comm(lbann_comm& comm);
```
Comment (benson31) [inline include/lbann/weights/weights.hpp]: This should probably take a `shared_ptr`.

File: src/weights/weights.cpp [hunk 4]
```diff
@@ -62,28 +62,23 @@ std::string get_dims_string(const std::vector<int>& matrix_height_dims,
...
-  m_matrix_dist.blockHeight = 1;
-  m_matrix_dist.blockWidth = 1;
-  m_matrix_dist.colAlign = 0;
-  m_matrix_dist.rowAlign = 0;
-  m_matrix_dist.colCut = 0;
-  m_matrix_dist.rowCut = 0;
-  m_matrix_dist.root = 0;
-  m_matrix_dist.grid = &(comm->get_trainer_grid());
-  m_matrix_dist.device = El::Device::CPU;
+weights::weights(lbann_comm* comm)
```
Comment (benson31) [inline src/weights/weights.cpp]: Again this should probably be `shared_ptr`.

File: src/weights/weights.cpp [hunk 5]
```diff
@@ -254,6 +249,25 @@ void weights::set_matrix_distribution(El::DistData dist) {
   m_matrix_dist = dist;
 }
 
+void weights::set_comm(lbann_comm& comm) {
```
Comment (benson31) [inline src/weights/weights.cpp]: Two things: 
1. This should probably be a `shared_ptr`.
2. Should this update the matrix distribution?

File: include/lbann/weights/weights.hpp (hunk 3 above)
Comment (bvanessen) [reply include/lbann/weights/weights.hpp]: Lets defer this.

File: src/weights/weights.cpp (hunk 4 above)
Comment (bvanessen) [reply src/weights/weights.cpp]: Deferred to future PR.

**Pull request merged**